tests/testrom.c

```diff
@@ -0,0 +1,246 @@
+	if (!(cond)) { \
+		LOG("ASSERTION FAILED:\n"); \
+		LOG("%s\n", #cond); \
+		LOG(msg, ##__VA_ARGS__); \
+		ctx->result = TEST_FAILED; \
+		return; \
+	} \
+})
+
+// ASSERT_EQUAL_HEX(a, b, msg): fail the test if a!=b (and log a & b as hex values)
+#define ASSERT_EQUAL_HEX(_a, _b, msg, ...) ({ \
+	uint64_t a = _a; uint64_t b = _b; \
+	if (a != b) { \
+		LOG("ASSERTION FAILED:\n"); \
+		LOG("%s != %s (0x%llx != 0x%llx)\n", #_a, #_b, a, b); \
+		LOG(msg, ## __VA_ARGS__); \
+		ctx->result = TEST_FAILED; \
+		return; \
+	} \
+})
+
+// ASSERT_EQUAL_UNSIGNED(a,b, msg): fail the test if a!=b (and log a & b as
+// unsigned values)
+#define ASSERT_EQUAL_UNSIGNED(_a, _b, msg, ...) ({ \
+	uint64_t a = _a; uint64_t b = _b; \
+	if (a != b) { \
+		LOG("ASSERTION FAILED:\n"); \
+		LOG("%s != %s (%llu != %llu)\n", #_a, #_b, a, b); \
+		LOG(msg, ## __VA_ARGS__); \
+		ctx->result = TEST_FAILED; \
+		return; \
+	} \
+})
+
+// ASSERT_EQUAL_SIGNED(a, b, msg): fail the test if a!=b (and log a/b as signed values)
+#define ASSERT_EQUAL_SIGNED(_a, _b, msg, ...) ({ \
+	int64_t a = _a; int64_t b = _b; \
+	if (a != b) { \
+		LOG("ASSERTION FAILED:\n"); \
+		LOG("%s != %s (%lld != %lld)\n", #_a, #_b, a, b); \
+		LOG(msg, ## __VA_ARGS__); \
+		ctx->result = TEST_FAILED; \
+		return; \
+	} \
+})
+
+void hexdump(char *out, const uint8_t *buf, int buflen, int start, int count) {
+	for (int i=start;i<start+count;i++) {
+		if (i >= 0 && i < buflen) {
+			sprintf(out, "%02x", buf[i]);
+			out += 2;
+		} else {
+			*out++ = '-'; *out++ = '-';
+		}
+	}	
+	*out = '\0';
+}
+
+int assert_equal_mem(TestContext *ctx, const uint8_t *a, const uint8_t *b, int len) {
+	for (int i=0;i<len;i++) {
+		if (a[i] != b[i]) {
+			char dumpa[64];
+			char dumpb[64];
+			hexdump(dumpa, a, len, i-2, 5);
+			hexdump(dumpb, b, len, i-2, 5);
+
+			LOG("ASSERTION FAILED:\n");
+			LOG("[%s] != [%s]\n", dumpa, dumpb);
+			LOG("     ^^              ^^  idx: %d\n", i);
+			return 0;
+		}
+	}
+	return 1;
+}
+
+// ASSERT_EQUAL_MEM(a, b, len, msg): fail the test if the memory pointer by
+// a and b has not the same content (up to len bytes).
+#define ASSERT_EQUAL_MEM(_a, _b, _len, msg, ...) ({ \
+	const uint8_t *a = (_a); const uint8_t *b = (_b); int len = (_len); \
+	if (!assert_equal_mem(ctx, a, b, len)) { \
+		LOG(msg, ## __VA_ARGS__); \
+		ctx->result = TEST_FAILED; \
+		return; \
+	} \
+})
+
+/**********************************************************************
+ * TEST FILES
+ **********************************************************************/
+
+#include "test_dfs.c"
+#include "test_cache.c"
+
+
+/**********************************************************************
+ * MAIN
+ **********************************************************************/
+
+// Testsuite definition
+#define TEST_FUNC(fn, dur)   { #fn, fn, dur }
+static const struct Testsuite
+{
+	const char *name;
+	TestFunc fn;
+	uint32_t duration;
+} tests[] = {
+	TEST_FUNC(test_dfs_read, 1104),
+	TEST_FUNC(test_cache_invalidate, 1337),
```

Comment:
I see what you did there.

Reply:
I won't try to make you believe that it was random... :)